espresso/streamer_test.go

```diff
@@ -0,0 +1,589 @@
+		}
+
+		state.AdvanceSafeL2()
+		state.AdvanceFinalizedL1()
+	}
+
+	if have, want := len(state.EspTransactionData), N; have != want {
+		t.Fatalf("unexpected number of batches in state:\nhave:\n\t%v\nwant:\n\t%v\n", have, want)
+	}
+}
+
+// TestStreamerEspressoOutOfOrder tests the behavior of the EspressoStreamer
+// when the batches coming from Espresso are not in sequential order.
+//
+// The Streamer is expected to be able to reorder these batches before
+// iterating over them.
+func TestStreamerEspressoOutOfOrder(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	namespace := uint64(42)
+	chainID := big.NewInt(int64(namespace))
+	privateKeyString := "59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d"
+	chainSignerFactory, signerAddress, _ := crypto.ChainSignerFactoryFromConfig(&NoOpLogger{}, privateKeyString, "", "", opsigner.CLIConfig{})
+	chainSigner := chainSignerFactory(chainID, common.Address{})
+
+	state, streamer := setupStreamerTesting(namespace, signerAddress)
+	rng := rand.New(rand.NewSource(0))
+
+	// update the state of our streamer
+	syncStatus := state.SyncStatus()
+	_, err := streamer.Refresh(ctx, syncStatus.FinalizedL1, syncStatus.SafeL2.Number)
+
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("failed to refresh streamer state encountered error:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	const N = 1000
+	var batches []*derive.SingularBatch
+	for i := 0; i < N; i++ {
+		batch, _, _, block := state.CreateEspressoTxnData(
+			ctx,
+			namespace,
+			rng,
+			chainID,
+			uint64(i)+1,
+			chainSigner,
+		)
+
+		rollEspBlockNumber := rng.Intn(N * 5)
+		for {
+			_, ok := state.EspTransactionData[BlockAndNamespace(uint64(rollEspBlockNumber), namespace)]
+			if ok {
+				// re-roll, if already populated.
+				rollEspBlockNumber = rng.Intn(N * 5)
+				continue
+			}
+
+			break
+		}
+
+		state.AddEspressoTransactionData(uint64(rollEspBlockNumber), namespace, block)
+		batches = append(batches, batch)
+	}
+
+	{
+
+		for i := 0; i < N; i++ {
+			if !streamer.HasNext(ctx) {
+				// Update the state of our streamer
+				if have, want := streamer.Update(ctx), error(nil); !errors.Is(have, want) {
+					t.Fatalf("failed to update streamer state encountered error:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+				}
+			}
+
+			batch := batches[i]
+			batchFromEsp := streamer.Next(ctx)
+			if have, want := batchFromEsp, (*derive.EspressoBatch)(nil); have == want {
+				t.Fatalf("unexpectedly did not received batch from streamer:\nhave:\n\t%v\ndo not want:\n\t%v\n", have, want)
+			}
+
+			// This batch ** should ** match the one we created above.
+
+			if have, want := batchFromEsp.Batch.GetEpochNum(), batch.GetEpochNum(); have != want {
+				t.Fatalf("batch epoch number does not match:\nhave:\n\t%v\ndo not want:\n\t%v\n", have, want)
+			}
+
+			state.AdvanceSafeL2()
+		}
```

Comment:
Any reason why we don't advance the finalized L1 head? 
```
state.AdvanceFinalizedL1()
```

Reply:
No particular reason, other than it seems like there isn't any direct need to in order to verify behavior.  We can add another test to advance the l1 to see how it affects things as well.  However, in order for it to have any affect on the `EspressoStreamer`, we would need to be calling `.Refresh` within the loop some where.